google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -75,68 +76,119 @@ Instant instant() {
    * Wrapper around {@code ReadContext} that releases the session to the pool once the call is
    * finished, if it is a single use context.
    */
-  private static class AutoClosingReadContext implements ReadContext {
-    private final ReadContext delegate;
+  private static class AutoClosingReadContext<T extends ReadContext> implements ReadContext {
+    private final Supplier<T> readContextDelegateSupplier;
+    private T readContextDelegate;
     private final PooledSession session;
     private final boolean isSingleUse;
     private boolean closed;
 
     private AutoClosingReadContext(
-        ReadContext delegate, PooledSession session, boolean isSingleUse) {
-      this.delegate = delegate;
+        Supplier<T> delegateSupplier, PooledSession session, boolean isSingleUse) {
+      this.readContextDelegateSupplier = delegateSupplier;
       this.session = session;
       this.isSingleUse = isSingleUse;
+      try {
+        this.readContextDelegate = delegateSupplier.get();
+      } catch (SessionNotFoundException e) {
+        recreateSessionIfPossible(e);
+      }
     }
 
-    private ResultSet wrap(final ResultSet resultSet) {
-      session.markUsed();
-      if (!isSingleUse) {
-        return resultSet;
+    T getReadContextDelegate() {
+      return readContextDelegate;
+    }
+
+    private ResultSet wrap(final Supplier<ResultSet> resultSetSupplier) {
+      ResultSet res;
+      try {
+        res = resultSetSupplier.get();
+      } catch (SessionNotFoundException e) {
+        recreateSessionIfPossible(e);
+        res = resultSetSupplier.get();
       }
-      return new ForwardingResultSet(resultSet) {
+      return new ForwardingResultSet(res) {
+        private boolean beforeFirst = true;
+
         @Override
         public boolean next() throws SpannerException {
-          try {
-            boolean ret = super.next();
-            if (!ret) {
-              close();
-            }
-            return ret;
-          } catch (SpannerException e) {
-            if (!closed) {
-              session.lastException = e;
-              AutoClosingReadContext.this.close();
+          while (true) {
+            try {
+              boolean ret = super.next();
+              if (beforeFirst) {
+                session.markUsed();
+                beforeFirst = false;
+              }
+              if (!ret && isSingleUse) {
+                close();
+              }
+              return ret;
+            } catch (SessionNotFoundException e) {
+              recreateSessionIfPossible(e);
+              replaceDelegate(resultSetSupplier.get());
```

Comment:
can we just do a `return next()` here instead of adding `while(true)`?

Reply:
I replaced the while(true) loop with an internal method that is can be re-used in the SessionNotFound handler.

Reply:
Should the while(true) still be here?